e3sm_diags/driver/cosp_histogram_driver.py

```diff
@@ -1,130 +1,145 @@
+                # variable with its spatial average.
+                ds_test_avg = ds_test.copy()
+                ds_ref_avg = ds_test.copy()
+                ds_test_avg[var_key] = spatial_avg(
+                    ds_test_region, var_key, as_list=False
                 )
-                utils.general.save_ncfiles(
-                    parameter.current_set,
-                    mv1_domain,
-                    mv2_domain,
-                    diff,
+                ds_ref_avg[var_key] = spatial_avg(ds_ref_region, var_key, as_list=False)
+
+                # The dimension names of both variables must be aligned to
+                # perform arithmetic with Xarray. Sometimes the dimension names
+                # might differ based on the derived variable (e.g.,
+                # "cosp_htmisr" vs. "misr_cth").
+                ds_test_avg[var_key] = _align_test_to_ref_dims(
+                    ds_test_avg[var_key], ds_ref_avg[var_key]
+                )
+                ds_diff_avg = _get_diff_of_avg(var_key, ds_test_avg, ds_ref_avg)
+
+                _save_data_metrics_and_plots(
                     parameter,
+                    plot_func,
+                    var_key,
+                    ds_test_avg,
+                    ds_ref_avg,
+                    ds_diff_avg,
+                    metrics_dict=None,
                 )
 
     return parameter
+
+
+def _get_diff_of_avg(
+    var_key: str, ds_test_avg: xr.Dataset, ds_ref_avg: xr.Dataset
+) -> xr.Dataset:
+    # Use the test dataset as the base dataset to subtract the reference dataset
+    # from.
+    ds_diff_avg = ds_test_avg.copy()
+
+    # There are case where the axes of the test and ref datasets aren't in the
+    # same units. We avoid label-based Xarray arithmetic which expect coordinates
+    # with the same units and will produce np.nan results by subtracting.
+    # Instead, we subtract using the reference xr.DataArray's `np.array`
+    # (`.values`).
+    ds_diff_avg[var_key] = ds_diff_avg[var_key] - ds_ref_avg[var_key].values
+
+    return ds_diff_avg
+
+
+def _align_test_to_ref_dims(
+    da_test: xr.DataArray, da_ref: xr.DataArray
+) -> xr.DataArray:
+    """Align the dimensions of the test data to the ref data.
+
+    This is useful for situations where label-based arithmetic needs to be
+    performed using Xarray.
+
+    Parameters
+    ----------
+    da_test : xr.DataArray
+        The test dataarray.
+    da_ref : xr.DataArray
+        The ref dataarray.
+
+    Returns
+    -------
+    xr.DataArray
+        The test dataarray with dimensions aligned to the ref dataarray.
+    """
+    da_test_new = da_test.copy()
+
+    # NOTE: This logic assumes that prs and tau are in the same order for
+    # the test and ref variables. If they are not, then this will break or
+    # perform incorrect arithmetic.
+    da_test_new = da_test_new.rename(
+        {dim1: dim2 for dim1, dim2 in zip(da_test.dims, da_ref.dims)}
+    )
```

Comment:
Notable comment

Reply:
Based on experience, I think it is the case.